# Appartment-Hunter

### Imports de bibliothèques

In [1]:
import pandas as pd
import os
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("data/raw/kc_house_data.csv")
df.head()

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,7129300520,20141013T000000,221900.0,3,1.00,1180,5650,1.0,0,0,...,7,1180,0,1955,0,98178,47.5112,-122.257,1340,5650
1,6414100192,20141209T000000,538000.0,3,2.25,2570,7242,2.0,0,0,...,7,2170,400,1951,1991,98125,47.7210,-122.319,1690,7639
2,5631500400,20150225T000000,180000.0,2,1.00,770,10000,1.0,0,0,...,6,770,0,1933,0,98028,47.7379,-122.233,2720,8062
3,2487200875,20141209T000000,604000.0,4,3.00,1960,5000,1.0,0,0,...,7,1050,910,1965,0,98136,47.5208,-122.393,1360,5000
4,1954400510,20150218T000000,510000.0,3,2.00,1680,8080,1.0,0,0,...,8,1680,0,1987,0,98074,47.6168,-122.045,1800,7503


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  str    
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long           21

In [4]:
print(df.isnull().sum().sum())

0


**Missing values: none** (0 across all 21 columns). There is nothing to impute or remove.

We find that the *date* column is an object type.
So in order to use it we'll need to turn it into a datetime type

In [5]:
df["date"]=pd.to_datetime(df["date"], errors='coerce')
df["date"].head()

0   2014-10-13
1   2014-12-09
2   2015-02-25
3   2014-12-09
4   2015-02-18
Name: date, dtype: datetime64[us]

Actually, the date is between 2014 and 2015 so the impact of this column is very low. So we'll drop it

In [6]:
df=df.drop("date", axis=1)

In [7]:
df.duplicated().sum()
df[df.duplicated()]

,id,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
3951,1825069031,550000.0,4,1.75,2410,8447,2.0,0,3,4,8,2060,350,1936,1980,98074,47.6499,-122.088,2520,14789
14983,6308000010,585000.0,3,2.50,2290,5089,2.0,0,0,3,9,2290,0,2001,0,98006,47.5443,-122.172,2290,7984
20054,8648900110,555000.0,3,2.50,1940,3211,2.0,0,0,3,8,1940,0,2009,0,98027,47.5644,-122.093,1880,3078


In [8]:
df = df.drop_duplicates()

**3 duplicates detected and removed.** Once the date is dropped, these are identical rows (same `id`, same price, same features): the same sale is counted twice. Keeping them would give these houses a double weight during training.

In [9]:
df.describe()

,id,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
count,2.161000e+04,2.161000e+04,21610.000000,21610.000000,21610.000000,2.161000e+04,21610.000000,21610.000000,21610.000000,21610.000000,21610.000000,21610.000000,21610.000000,21610.000000,21610.000000,21610.000000,21610.000000,21610.000000,21610.000000,21610.000000
mean,4.580161e+09,5.400849e+05,3.370847,2.114739,2079.881212,1.510829e+04,1.494239,0.007543,0.234197,3.409440,7.656779,1788.347894,291.533318,1971.003609,84.322351,98077.945673,47.560049,-122.213910,1986.518695,12769.031976
std,2.876547e+09,3.671525e+05,0.930110,0.770204,918.500299,4.142323e+04,0.539994,0.086523,0.766136,0.650764,1.175500,828.138723,442.596699,29.372639,401.499264,53.505373,0.138572,0.140833,685.425781,27305.972464
min,1.000102e+06,7.500000e+04,0.000000,0.000000,290.000000,5.200000e+02,1.000000,0.000000,0.000000,1.000000,1.000000,290.000000,0.000000,1900.000000,0.000000,98001.000000,47.155900,-122.519000,399.000000,651.000000
25%,2.123049e+09,3.216125e+05,3.000000,1.750000,1425.500000,5.040000e+03,1.000000,0.000000,0.000000,3.000000,7.000000,1190.000000,0.000000,1951.000000,0.000000,98033.000000,47.470925,-122.328000,1490.000000,5100.000000
50%,3.904930e+09,4.500000e+05,3.000000,2.250000,1910.000000,7.619000e+03,1.500000,0.000000,0.000000,3.000000,7.000000,1560.000000,0.000000,1975.000000,0.000000,98065.000000,47.571800,-122.231000,1840.000000,7620.000000
75%,7.308900e+09,6.450000e+05,4.000000,2.500000,2550.000000,1.068875e+04,2.000000,0.000000,0.000000,4.000000,8.000000,2210.000000,560.000000,1997.000000,0.000000,98118.000000,47.678000,-122.125000,2360.000000,10083.000000
max,9.900000e+09,7.700000e+06,33.000000,8.000000,13540.000000,1.651359e+06,3.500000,1.000000,4.000000,5.000000,13.000000,9410.000000,4820.000000,2015.000000,2015.000000,98199.000000,47.777600,-121.315000,6210.000000,871200.000000


In [10]:
aberrant = df[(df['yr_renovated'] != 0) & (df['yr_renovated'] < df['yr_built'])]
print(len(aberrant))

0


No houses have been renovated before the year date. So these columns seem ok to keep.

In [11]:
# Houses with many more bathrooms than bedrooms
rooms_outlier = df[(df["bathrooms"] > df["bedrooms"] + 2)]
print(len(rooms_outlier))
rooms_outlier[["price", "bedrooms", "bathrooms", "sqft_living"]]

3


,price,bedrooms,bathrooms,sqft_living
8477,339950.0,0,2.5,2290
8484,240000.0,0,2.5,1810
12653,320000.0,0,2.5,1490


**Decision: remove the inconsistent rows.**

The 3 detected houses have 0 bedrooms for 2.5 bathrooms and 1,490 to 2,290 sqft: this is a data entry error, not a real home. We also check the maximum number of bedrooms: a house with 33 bedrooms for 1,620 sqft and 1.75 bathrooms is a typo (probably 3).

We remove these 4 rows out of 21,613 (0.02%). The loss of information is negligible, and the wrong values would distort the relation between bedrooms and price. We keep the other houses with 8-11 bedrooms: their living area is consistent.

In [12]:
# House with 33 bedrooms for 1620 sqft: data entry error
bedrooms_error = df[df["bedrooms"] == 33]
print(len(bedrooms_error))

# Remove the inconsistent rows, then the id column (an identifier with no predictive value)
df = df.drop(rooms_outlier.index)
df = df.drop(bedrooms_error.index)
df = df.drop('id', axis=1)
print(df.shape)

1
(21606, 19)


In [13]:
df["price_per_living_sqft"]=df["price"]/df['sqft_living']
df["price_per_lot_sqft"]=df["price"]/df['sqft_lot']

In [14]:
# Areas: sqft -> sqm (1 sqft = 0.092903 m²)
df['sqm_living'] = df['sqft_living'] * 0.092903
df['sqm_lot'] = df['sqft_lot'] * 0.092903
df['sqm_basement'] = df['sqft_basement'] * 0.092903
df['sqm_living15'] = df['sqft_living15'] * 0.092903
df['sqm_lot15'] = df['sqft_lot15'] * 0.092903
df['sqm_above'] = df['sqft_above'] * 0.092903

# Price per sqm: DIVIDE the price per sqft by 0.092903 (1 m² = 10.76 sqft)
df['sqm_price_per_lot'] = df['price_per_lot_sqft'] / 0.092903
df['sqm_price_per_living'] = df['price_per_living_sqft'] / 0.092903

# Drop the sqft columns: they are replaced by the sqm columns
sqft_columns = ['sqft_living', 'sqft_lot', 'sqft_basement', 'sqft_living15',
                'sqft_lot15', 'sqft_above', 'price_per_living_sqft', 'price_per_lot_sqft']
df = df.drop(columns=sqft_columns)

### Data distribution for visualizing potential outliers

In [15]:
fig, axes = plt.subplots(2, 3, figsize=(16, 9))
fig.suptitle("Distribution of categorical variables", fontsize=14)

categoricals = ['bedrooms', 'floors', 'waterfront', 'view', 'condition', 'grade']
titles = ['Bedrooms', 'Floors', 'Waterfront (0/1)', 'View (0-4)', 'Condition (1-5)', 'Grade (1-13)']

for ax, col, title in zip(axes.flatten(), categoricals, titles):
    counts = df[col].value_counts().sort_index()
    ax.bar(range(len(counts)), counts.values, color='steelblue', edgecolor='black')
    ax.set_title(title)
    ax.set_xticks(range(len(counts)))
    ax.set_xticklabels([str(i) for i in counts.index], rotation=0)
plt.tight_layout()
plt.show()



fig, axes = plt.subplots(1, 5, figsize=(18, 5))
fig.suptitle("Distribution of continuous variables", fontsize=14)
for ax, col, color in zip(axes,
                           ['sqm_living', 'sqm_lot', 'sqm_above', 'yr_built', 'yr_renovated'],
                           ['cornflowerblue', 'mediumpurple', 'coral', 'tomato', 'tomato']):
    ax.hist(df[col], bins=40, color=color, edgecolor='black', alpha=0.8)
    ax.set_title(col)
plt.tight_layout()
plt.show()



plt.figure(figsize=(5, 4))
plt.hist(df['price'], bins=60, color='#2ecc71', edgecolor='black')
plt.title("Distribution of the target variable")
plt.xlabel("Price ($)")
plt.ylabel("Number of houses")
plt.tight_layout()
plt.show()
print(f"Price statistics : {df['price'].describe().round(0).to_dict()}")

C:\Users\kylli\AppData\Local\Temp\ipykernel_4156\1791231763.py:14: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()
C:\Users\kylli\AppData\Local\Temp\ipykernel_4156\1791231763.py:26: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()
C:\Users\kylli\AppData\Local\Temp\ipykernel_4156\1791231763.py:36: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


Price statistics : {'count': 21606.0, 'mean': 540114.0, 'std': 367175.0, 'min': 75000.0, '25%': 321962.0, '50%': 450000.0, '75%': 645000.0, 'max': 7700000.0}


In [16]:
fig, axes = plt.subplots(1, 5, figsize=(20, 5))
fig.suptitle("Box plots of continuous variables", fontsize=14)
for ax, col in zip(axes, ['price', 'sqm_living', 'sqm_lot', 'bedrooms', 'bathrooms']):
    ax.boxplot(df[col], patch_artist=True,
               boxprops=dict(facecolor='lightblue', color='navy'),
               medianprops=dict(color='red', linewidth=2))
    ax.set_title(col)
plt.tight_layout()
plt.show()


fig, axes = plt.subplots(1, 5, figsize=(20, 5))
fig.suptitle("Continuous variables by price (below / above median)",
             fontsize=13)
for ax, col in zip(axes, ['sqm_living', 'grade', 'bathrooms', 'sqm_above', 'sqm_price_per_living']):
    for val, label, color in [(False, 'Price < median', '#2ecc71'), (True, 'Price ≥ median', '#e74c3c')]:
        ax.hist(df[(df['price'] >= df['price'].median()) == val][col], bins=40,
                alpha=0.6, label=label, color=color, density=True)
    ax.set_title(col)
    ax.legend()
plt.tight_layout()
plt.show()

C:\Users\kylli\AppData\Local\Temp\ipykernel_4156\2403575121.py:9: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


C:\Users\kylli\AppData\Local\Temp\ipykernel_4156\2403575121.py:22: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [17]:
price = df['price']

Q1 = price.quantile(0.25)
Q3 = price.quantile(0.75)
IQR = Q3 - Q1
iqr_mask = (price < Q1 - 1.5 * IQR) | (price > Q3 + 1.5 * IQR)
print(f"Outliers IQR : {iqr_mask.sum()}")

median = price.median()
mad = (price - median).abs().median()
modified_z = 0.6745 * (price - median) / mad
mad_mask = abs(modified_z) > 3.5
print(f"Outliers modified Z-score : {mad_mask.sum()}")

low = price.quantile(0.01)
high = price.quantile(0.99)
pct_mask = (price < low) | (price > high)
print(f"Outliers percentiles : {pct_mask.sum()}")

outliers_combined = df[iqr_mask & mad_mask & pct_mask]
print(f"Outliers in the 3 methods : {len(outliers_combined)}")
outliers_combined[['price','sqm_living','grade','bedrooms']].describe()

Outliers IQR : 1146
Outliers modified Z-score : 928
Outliers percentiles : 434
Outliers in the 3 methods : 217


,price,sqm_living,grade,bedrooms
count,2.170000e+02,217.000000,217.000000,217.000000
mean,2.706587e+06,462.613699,10.677419,4.350230
std,8.368498e+05,143.016575,1.141432,1.016946
min,1.965000e+06,242.476830,7.000000,2.000000
25%,2.205000e+06,375.328120,10.000000,4.000000
50%,2.453500e+06,427.353800,11.000000,4.000000
75%,2.950000e+06,509.108440,11.000000,5.000000
max,7.700000e+06,1257.906620,13.000000,8.000000


**Decision: we keep the 217 price outliers.**

- These are the houses flagged by all 3 methods at once: prices from 1.96 M$ to over 7 M$, large areas (463 m² on average) and a high grade (7 to 13). They are consistent with each other: they are real luxury homes, not errors.
- Removing them would remove the whole top of the market: the model would be unable to predict these properties.
- To limit their weight in the models, we will use `log(price)` as the target during modeling, and use models that are not very sensitive to extreme values (Random Forest, Gradient Boosting).
- The price stays in its original value in the cleaned file (the log is a modeling step, not a cleaning step).

**Standardization: not in this notebook.**

Standardization (`StandardScaler`) will be done inside the modeling pipeline, after the train / test split: the mean and standard deviation are computed on the train set only, otherwise information from the test set would leak into the training. It only concerns models sensitive to scale (SVR). `zipcode` will be treated as a categorical variable (One-Hot Encoding).

**To exclude from modeling:** `sqm_price_per_lot` and `sqm_price_per_living`. They are computed from the price and would give the answer to the model. They stay in the cleaned file for the exploratory analysis (Power BI).

### Saving in clean folder

In [18]:
os.makedirs("data/clean", exist_ok=True)

df.to_csv("data/clean/kc_house_data_clean.csv", index=False)

print("\nSaved in clean folder.")


Saved in clean folder.


## Feature selection

The goal is to choose the features the models will use. We compare four methods, because each one has its own point of view:

1. **Correlation** with the price (simple, but only linear).
2. **Random Forest importance** (captures non-linear links).
3. **Boruta** (keeps a feature only if it does better than random noise).
4. **Forward feature selection** (adds features one by one while the score improves).

Rules used for all the methods:
- The target is `log(price)`, as decided in the cleaning step.
- The selection is done on the **train set only**, so the test set stays untouched.
- `sqm_price_per_lot` and `sqm_price_per_living` are excluded (they are computed from the price).
- `zipcode` is left out of the four methods: it is a category, not a quantity, so a numeric code would mislead them. It is tested separately at the end.

In [19]:
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.ensemble import RandomForestRegressor
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from boruta import BorutaPy

In [20]:
# Target: log of the price (reduces the weight of the luxury houses)
y = np.log(df["price"])

# Features: every column except the price, the columns computed from the price
# and zipcode (zipcode is a category, it is tested separately at the end)
excluded = ["price", "sqm_price_per_lot", "sqm_price_per_living", "zipcode"]
features = [col for col in df.columns if col not in excluded]
X = df[features]

# The selection is done on the train set only, so the test set stays untouched
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(len(features), "features")
print(X_train.shape, X_test.shape)

17 features
(17284, 17) (4322, 17)


### 1. Correlation

`grade` (0.70) and `sqm_living` (0.69) are the most linked to the price, followed by `sqm_living15`, `sqm_above`, `bathrooms` and `lat`. `condition`, `long` and `yr_built` have almost no linear link.

On the heatmap, only `sqm_living` and `sqm_above` are strongly correlated (0.88): they carry almost the same information, so we only need one of them. Be careful, `long` has a low correlation but can still matter, because the price depends on the position (`lat` and `long` together).

In [21]:
# Correlation of each feature with the (log) price
corr_price = X_train.corrwith(y_train).sort_values(ascending=False)
print(corr_price.round(2))

plt.figure(figsize=(8, 5))
corr_price.plot(kind="barh", color="steelblue")
plt.gca().invert_yaxis()
plt.title("Correlation with log(price)")
plt.tight_layout()
plt.show()

grade           0.70
sqm_living      0.69
sqm_living15    0.62
sqm_above       0.60
bathrooms       0.55
lat             0.45
bedrooms        0.35
view            0.34
sqm_basement    0.31
floors          0.31
waterfront      0.16
yr_renovated    0.11
sqm_lot         0.10
sqm_lot15       0.09
yr_built        0.08
long            0.05
condition       0.05
dtype: float64


C:\Users\kylli\AppData\Local\Temp\ipykernel_4156\1911194493.py:10: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [22]:
# Correlation between the features (to find redundant ones)
corr_features = X_train.corr()

plt.figure(figsize=(12, 9))
sns.heatmap(corr_features, cmap="coolwarm", center=0, annot=True, fmt=".1f", annot_kws={"size": 7})
plt.title("Correlation between features")
plt.tight_layout()
plt.show()

# Pairs of features with a correlation above 0.8
for i in range(len(features)):
    for j in range(i + 1, len(features)):
        if abs(corr_features.iloc[i, j]) > 0.8:
            print(features[i], "/", features[j], ":", round(corr_features.iloc[i, j], 2))

sqm_living / sqm_above : 0.88


C:\Users\kylli\AppData\Local\Temp\ipykernel_4156\1193659830.py:8: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


### 2. Random Forest importance

The importance is concentrated on three features: `grade`, `lat` and `sqm_living` (about 82% of the total). Features that are correlated share their importance (for example `sqm_above` with `sqm_living`), so a low importance does not always mean a useless feature. We keep the features above the average importance.

In [23]:
# Random Forest feature importance
rf = RandomForestRegressor(n_estimators=200, random_state=42, n_jobs=-1)
rf.fit(X_train, y_train)

rf_importance = pd.Series(rf.feature_importances_, index=features).sort_values(ascending=False)
print(rf_importance.round(3))

plt.figure(figsize=(8, 5))
rf_importance.plot(kind="barh", color="seagreen")
plt.gca().invert_yaxis()
plt.title("Random Forest feature importance")
plt.tight_layout()
plt.show()

# We keep the features above the average importance
rf_selected = rf_importance[rf_importance > rf_importance.mean()].index.tolist()
print(rf_selected)

grade           0.344
lat             0.295
sqm_living      0.181
long            0.049
sqm_living15    0.027
yr_built        0.022
sqm_lot         0.017
sqm_lot15       0.015
sqm_above       0.014
view            0.010
waterfront      0.006
bathrooms       0.006
condition       0.005
sqm_basement    0.003
bedrooms        0.003
floors          0.002
yr_renovated    0.001
dtype: float64
['grade', 'lat', 'sqm_living']


C:\Users\kylli\AppData\Local\Temp\ipykernel_4156\3071745326.py:13: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


### 3. Boruta

Boruta is the most permissive method: it keeps 12 features out of 17, every feature that does better than shuffled noise. It rejects `bedrooms`, `floors`, `sqm_basement`, `yr_renovated` and `condition`. No feature is left undecided (no tentative one).

In [24]:
# Boruta: compares each feature to shuffled copies of the features
rf_boruta = RandomForestRegressor(n_jobs=-1, max_depth=7, random_state=42)
boruta = BorutaPy(rf_boruta, n_estimators="auto", max_iter=50, random_state=42)
boruta.fit(X_train.values, y_train.values)

boruta_selected = [f for f, keep in zip(features, boruta.support_) if keep]
boruta_tentative = [f for f, maybe in zip(features, boruta.support_weak_) if maybe]
print("Confirmed :", boruta_selected)
print("Tentative :", boruta_tentative)

Confirmed : ['bathrooms', 'waterfront', 'view', 'grade', 'yr_built', 'lat', 'long', 'sqm_living', 'sqm_lot', 'sqm_living15', 'sqm_lot15', 'sqm_above']
Tentative : []


### 4. Forward feature selection

This method is the most compact: it stops when a new feature improves the R2 by less than 0.001, and keeps 7 features. It avoids redundancy, because a feature that brings the same information as one already chosen is not added.

In [25]:
# Forward feature selection: we add the feature that improves the R2 the most, one by one
rf_forward = RandomForestRegressor(n_estimators=50, max_depth=10, random_state=42, n_jobs=-1)
forward = SequentialFeatureSelector(rf_forward, n_features_to_select="auto", tol=0.001,
                                    direction="forward", scoring="r2", cv=3)
forward.fit(X_train, y_train)

forward_selected = [f for f, keep in zip(features, forward.get_support()) if keep]
print(forward_selected)

['waterfront', 'view', 'grade', 'lat', 'long', 'sqm_living', 'sqm_lot']


### Summary of the four methods

Each method votes for the features it keeps. Four votes means every method agrees.

In [26]:
# Summary: one column per method, True if the method keeps the feature
summary = pd.DataFrame(index=features)
summary["correlation"] = corr_price.abs() >= 0.1
summary["random_forest"] = [f in rf_selected for f in features]
summary["boruta"] = [f in boruta_selected for f in features]
summary["forward"] = [f in forward_selected for f in features]
summary["votes"] = summary.sum(axis=1)
summary.sort_values("votes", ascending=False)

,correlation,random_forest,boruta,forward,votes
grade,True,True,True,True,4
sqm_living,True,True,True,True,4
lat,True,True,True,True,4
waterfront,True,False,True,True,3
view,True,False,True,True,3
sqm_lot,True,False,True,True,3
sqm_living15,True,False,True,False,2
long,False,False,True,True,2
bathrooms,True,False,True,False,2
sqm_above,True,False,True,False,2


### Final choice

We keep the features chosen by **at least 3 methods out of 4**, and we add `long`:
- `long` has only 2 votes (Boruta and forward selection), but a position needs both coordinates. `lat` alone gives the north-south axis only.
- The features dropped are either redundant (`sqm_above` is almost the same as `sqm_living`, correlation of 0.88), or their information is already in a kept feature (`bathrooms` and `sqm_living15` with `sqm_living` and `grade`), or they bring almost nothing (`condition`, `floors`, `yr_renovated`, `bedrooms`).

We check this choice below: a good selection must keep the score of the model, with fewer features.

In [27]:
# Final features: at least 3 votes out of 4, plus long (a position needs lat and long)
final_features = summary[summary["votes"] >= 3].index.tolist()
final_features.append("long")
print(final_features)

['waterfront', 'view', 'grade', 'lat', 'sqm_living', 'sqm_lot', 'long']


In [28]:
# Check: R2 of a Random Forest in cross-validation, with all the features and with the final ones
rf_check = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)

feature_sets = {
    "all features (17)": features,
    "final features": final_features,
    "final features without long": [f for f in final_features if f != "long"],
}

for name, cols in feature_sets.items():
    scores = cross_val_score(rf_check, X_train[cols], y_train, cv=3, scoring="r2")
    print(name, ":", round(scores.mean(), 4))

all features (17) : 0.8805


final features : 0.8771


final features without long : 0.8337


**Check result** (R2 on the log of the price, cross-validation on the train set):

| Features | R2 |
| --- | --- |
| All 17 features | 0.8805 |
| 7 final features | 0.8771 |
| 6 features (without `long`) | 0.8337 |

With less than half of the features, the score drops by only 0.003. Removing `long` costs 0.04, which confirms that we were right to keep it.

### Location: lat/long or lat/long + zipcode

Two configurations are compared with the same model: model 1 uses `lat` and `long`, model 2 adds `zipcode` (One-Hot Encoding, because a zipcode is a category).

In [29]:
# Model 1: final features (with lat and long)
X1 = X_train[final_features]

# Model 2: the same features + zipcode, encoded with One-Hot Encoding
X2 = X1.join(df.loc[X_train.index, ["zipcode"]])
encoder = ColumnTransformer([("zipcode", OneHotEncoder(handle_unknown="ignore"), ["zipcode"])],
                            remainder="passthrough")
model2 = Pipeline([("encoder", encoder),
                   ("forest", RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1))])

scores1 = cross_val_score(rf_check, X1, y_train, cv=3, scoring="r2")
scores2 = cross_val_score(model2, X2, y_train, cv=3, scoring="r2")
print("Model 1 (lat, long)          :", round(scores1.mean(), 4))
print("Model 2 (lat, long, zipcode) :", round(scores2.mean(), 4))

Model 1 (lat, long)          : 0.8771
Model 2 (lat, long, zipcode) : 0.8778


**Conclusion of the feature selection**

Adding `zipcode` gives 0.8778 instead of 0.8771: a gain of 0.0007, which is negligible. `lat` and `long` already describe the location well, so we keep **model 1** (without `zipcode`). It is simpler and avoids about 70 extra columns.

**Final features (7):** `grade`, `sqm_living`, `sqm_lot`, `waterfront`, `view`, `lat`, `long`.

They cover the three ideas found in the exploration: the **size** (`sqm_living`, `sqm_lot`), the **quality** (`grade`, `view`, `waterfront`) and the **location** (`lat`, `long`).